CodeU_Assignment_6/test/codeu_assignment_6/KemRearrangingCarsTest.java

```diff
@@ -0,0 +1,185 @@
+        
+        assertEquals(expectedRes, (new KemRearrangingCars()).getMovesRearrangingCars(permutationA, permutationB));
+    }
+    
+    @Test
+    public void testGetMovesRearrangingCars7Items0inMiddle() {
+        System.out.println("getMovesRearrangingCars 7 items with free space in the middle");
+        Integer[] permutationA = {1,5,4,0,2,3,6};
+        Integer[] permutationB = {0,1,2,3,4,5,6};
+        ArrayList<Pair<Integer, Integer>> expectedRes = new ArrayList<>();
+        expectedRes.add(new Pair<>(0,3));
+        expectedRes.add(new Pair<>(1,0));
+        expectedRes.add(new Pair<>(3,1));
+        expectedRes.add(new Pair<>(2,3));
+        expectedRes.add(new Pair<>(4,2));
+        expectedRes.add(new Pair<>(3,4));
+        expectedRes.add(new Pair<>(5,3));
+        expectedRes.add(new Pair<>(0,5));
+        
+        assertEquals(expectedRes, (new KemRearrangingCars()).getMovesRearrangingCars(permutationA, permutationB));
+    }
+    
+    @Test
+    public void testGetMovesRearrangingCars10ItemsMix() {
+        System.out.println("getMovesRearrangingCars 10 items, test 1");
+        Integer[] permutationA = {0,1,2,3,4,5,6,7,8,9};
+        Integer[] permutationB = {6,4,2,7,9,0,3,5,1,8};
+        ArrayList<Pair<Integer, Integer>> expectedRes = new ArrayList<>();
+        expectedRes.add(new Pair<>(6,0));
+        expectedRes.add(new Pair<>(1,6));
+        expectedRes.add(new Pair<>(4,1));
+        expectedRes.add(new Pair<>(3,4));
+        expectedRes.add(new Pair<>(7,3));
+        expectedRes.add(new Pair<>(4,7));
+        expectedRes.add(new Pair<>(9,4));
+        expectedRes.add(new Pair<>(5,9));
+        expectedRes.add(new Pair<>(6,5));
+        expectedRes.add(new Pair<>(7,6));
+        expectedRes.add(new Pair<>(9,7));
+        expectedRes.add(new Pair<>(8,9));
+        expectedRes.add(new Pair<>(5,8));
+        
+        assertEquals(expectedRes, (new KemRearrangingCars()).getMovesRearrangingCars(permutationA, permutationB));
+    }
+    
+    @Test
+    public void testGetMovesRearrangingCars10ItemsMixToMix() {
+        System.out.println("getMovesRearrangingCars 10 items, test 2");
+        Integer[] permutationA = {8,0,3,7,5,2,4,1,9,6};
+        Integer[] permutationB = {6,4,2,7,9,0,3,5,1,8};
+        ArrayList<Pair<Integer, Integer>> expectedRes = new ArrayList<>();;
+        expectedRes.add(new Pair<>(0,1)); 
+        expectedRes.add(new Pair<>(9,0));
+        expectedRes.add(new Pair<>(1,9));
+        expectedRes.add(new Pair<>(6,1));
+        expectedRes.add(new Pair<>(2,6)); 
+        expectedRes.add(new Pair<>(5,2)); 
+        expectedRes.add(new Pair<>(4,5));
+        expectedRes.add(new Pair<>(8,4)); 
+        expectedRes.add(new Pair<>(5,8));
+        expectedRes.add(new Pair<>(7,5));
+        expectedRes.add(new Pair<>(8,7));
+        expectedRes.add(new Pair<>(5,8));
+        assertEquals(expectedRes, (new KemRearrangingCars()).getMovesRearrangingCars(permutationA, permutationB));
+    }
+}

```

Comment:
The base/small input case coverage is good, and I see you cover a good range of input sizes.

A shorter way of building the expectedRes lists is described at https://stackoverflow.com/questions/1005073/initialization-of-an-arraylist-in-one-line - you can just use Arrays.asList.

A better way is to write a method that performs the sequence of moves, and checks that it achieves what you wanted. Why? There's more than one sequence of moves that will achieve the result, and your tests are brittle - they will fail if you run the test against someone else's algorithm. Brittle tests make a codebase difficult to maintain (you change the algorithm, it still produces a valid series of moves, but your tests break - annoying!).

Finally, when writing tests, try to concentrate on the code paths and corner cases you're trying to exercise - e.g. "does it work if numbers are swapped pairwise (2, 1, 4, 3, 6, 5, etc.)?", "does it work if zero is where it will end up?", "does it work if everything's already in the right place?", "does it work if everything is moved around cyclically by one place left/right?".

It looks to me like the tests you've written have good coverage, and cover most of these kinds of ideas, but it's worth making it explicit in your tests. Think "How could I have got the algorithm wrong, and what test input would discover that kind of mistake?".